liblwgeom/lwgeom_transform.c

```diff
@@ -172,3 +205,238 @@ lwproj_from_string(const char *str1)
 	}
 	return pj_init_plus(str1);
 }
+
+/***************************************************************************/
+
+#else /* POSTGIS_PROJ_VERION >= 60 */
+
+int
+lwgeom_transform_from_str(LWGEOM *geom, const char* instr, const char* outstr)
+{
+	PJ *pj = proj_create_crs_to_crs(NULL, instr, outstr, NULL);
+	if (!pj)
+	{
+		PJ *pj_in = proj_create(NULL, instr);
+		PJ *pj_out = proj_create(NULL, outstr);
+		if (!pj_in)
+			lwerror("could not parse proj string '%s'", instr);
+		if (!pj_out)
+			lwerror("could not parse proj string '%s'", outstr);
+		return LW_FAILURE;
+	}
+
+	return lwgeom_transform(geom, pj);
+}
+
+int
+lwgeom_transform(LWGEOM* geom, PJ* pj)
+{
+	uint32_t i;
+
+	/* No points to transform in an empty! */
+	if (lwgeom_is_empty(geom))
+		return LW_SUCCESS;
+
+	switch(geom->type)
+	{
+		case POINTTYPE:
+		case LINETYPE:
+		case CIRCSTRINGTYPE:
+		case TRIANGLETYPE:
+		{
+			LWLINE *g = (LWLINE*)geom;
+			if (!ptarray_transform(g->points, pj))
+				return LW_FAILURE;
+			break;
+		}
+		case POLYGONTYPE:
+		{
+			LWPOLY *g = (LWPOLY*)geom;
+			for (i = 0; i < g->nrings; i++)
+			{
+				if (!ptarray_transform(g->rings[i], pj))
+					return LW_FAILURE;
+			}
+			break;
+		}
+		case MULTIPOINTTYPE:
+		case MULTILINETYPE:
+		case MULTIPOLYGONTYPE:
+		case COLLECTIONTYPE:
+		case COMPOUNDTYPE:
+		case CURVEPOLYTYPE:
+		case MULTICURVETYPE:
+		case MULTISURFACETYPE:
+		case POLYHEDRALSURFACETYPE:
+		case TINTYPE:
+		{
+			LWCOLLECTION *g = (LWCOLLECTION*)geom;
+			for (i = 0; i < g->ngeoms; i++)
+			{
+				if (!lwgeom_transform(g->geoms[i], pj))
+					return LW_FAILURE;
+			}
+			break;
+		}
+		default:
+		{
+			lwerror("lwgeom_transform: Cannot handle type '%s'",
+			          lwtype_name(geom->type));
+			return LW_FAILURE;
+		}
+	}
+	return LW_SUCCESS;
+}
+
+static int
+proj_crs_is_swapped(const PJ* pj_crs)
+{
+	PJ *pj_cs = proj_crs_get_coordinate_system(NULL, pj_crs);
+	int axis_count = proj_cs_get_axis_count(NULL, pj_cs);
+	if (axis_count > 0)
+	{
+		const char *out_name, *out_abbrev, *out_direction;
+		double out_unit_conv_factor;
+		const char *out_unit_name, *out_unit_auth_name, *out_unit_code;
+		/* Read only first axis, see if it is degrees / north */
+		proj_cs_get_axis_info(NULL, pj_cs, 0,
+			&out_name,
+			&out_abbrev,
+			&out_direction,
+			&out_unit_conv_factor,
+			&out_unit_name,
+			&out_unit_auth_name,
+			&out_unit_code
+			);
+		return (strcasecmp(out_direction, "north") == 0);
+	}
+	return LW_FALSE;
+}
+
+
+int
+ptarray_transform(POINTARRAY* pa, PJ* pj)
+{
+	uint32_t i;
+	POINT4D p;
+	size_t n_converted;
+	size_t n_points = pa->npoints;
+	size_t point_size = ptarray_point_size(pa);
+	int has_z = ptarray_has_z(pa);
+	double *pa_double = (double*)(pa->serialized_pointlist);
+
+	/* XXXX TODO check that the PJ has decimal degrees as units in input/output */
+
+	int input_swapped = proj_crs_is_swapped(proj_get_source_crs(NULL, pj));
```

Comment:
I'm getting a coredump here while running the tests. `proj_get_source_crs` can return NULL and should be checked before calling `proj_get_source_crs`:
```
(gdb) bt
#0  0x00007f2beb28dd7f in raise () from /usr/lib/libc.so.6
#1  0x00007f2beb278672 in abort () from /usr/lib/libc.so.6
#2  0x00007f2beb278548 in __assert_fail_base.cold.0 () from /usr/lib/libc.so.6
#3  0x00007f2beb286396 in __assert_fail () from /usr/lib/libc.so.6
#4  0x00007f29da805062 in proj_crs_get_coordinate_system (ctx=0x7f29da8fa7c0 <pj_get_default_ctx::default_context>, crs=<optimized out>) at /usr/include/c++/8.2.1/ext/atomicity.h:69
#5  0x00007f29daa224e8 in proj_crs_is_swapped (pj_crs=0x0) at lwgeom_transform.c:294
#6  ptarray_transform (pa=0x55cf62e4b4f0, pj=0x55cf6365e7f0) at lwgeom_transform.c:330
#7  0x00007f29daa22359 in lwgeom_transform (geom=0x55cf62e4b538, pj=0x55cf6365e7f0) at lwgeom_transform.c:248
#8  0x00007f29da9754f7 in transform (fcinfo=<optimized out>) at lwgeom_transform.c:83
#9  0x000055cf627be130 in fmgr_security_definer (fcinfo=0x55cf62e495d0) at fmgr.c:748
#10 0x000055cf624e93e0 in ExecInterpExpr (state=<optimized out>, econtext=<optimized out>, isnull=0x7ffda3a9329f) at execExprInterp.c:678
#11 0x000055cf625b55ae in ExecEvalExprSwitchContext (state=<optimized out>, econtext=0x7ffda3a92bb0, isNull=0x87e417277d840000) at ../../../../src/include/executor/executor.h:303
#12 evaluate_expr (expr=<optimized out>, result_type=3353058, result_typmod=-1, result_collation=0) at clauses.c:4970
#13 0x000055cf625b4ddf in evaluate_function (funcid=3353401, result_type=<optimized out>, result_typmod=-1, result_collid=0, input_collid=0, args=0x55cf62f84900, context=<optimized out>, funcvariadic=<optimized out>, func_tuple=<optimized out>) at clauses.c:4512
#14 simplify_function (funcid=3353401, result_type=<optimized out>, result_typmod=-1, result_collid=0, input_collid=0, args_p=<optimized out>, funcvariadic=<optimized out>, process_args=<optimized out>, allow_non_const=true, context=0x7ffda3a937e0) at clauses.c:4152
#15 0x000055cf625b2fd1 in eval_const_expressions_mutator (node=<optimized out>, context=<optimized out>) at clauses.c:2742
```

Reply:
Also, according to the documentation of `proj_get_source_crs`:
```
 * The returned object must be unreferenced with proj_destroy() after
 * use.
```

So you are leaking memory here.

Reply:
I wonder what you're running in tests that gets you a NULL? Will see if I get an error in regress now that I'm guarding against that.

Reply:
It happened when running regress/core/ but I don't recall the exact test now. I can retry in a different PC tomorrow if you want (same OS and libraries, but different PC) or in that same machine on Friday if you need.

Reply:
This happened when running `../loader/ReprojectPts`, now it doesn't crash and simply fails. I think this is an issue with my environment (I wasn't packing the datumgrids in my custom packages), but test failures are ok, crashing wasn't.

Reply:
I can confirm that once I've rebuilt with current master and adding the datumgrids, the tests pass.